lib/SILOptimizer/Utils/Local.cpp

```diff
@@ -1981,9 +1981,88 @@ CastOptimizer::simplifyCheckedCastBranchInst(CheckedCastBranchInst *Inst) {
   return NewI;
 }
 
-SILInstruction *
-CastOptimizer::
-optimizeCheckedCastAddrBranchInst(CheckedCastAddrBranchInst *Inst) {
+SILInstruction *CastOptimizer::simplifyCheckedCastValueBranchInst(
+    CheckedCastValueBranchInst *Inst) {
+  if (auto *I = optimizeCheckedCastValueBranchInst(Inst))
+    Inst = dyn_cast<CheckedCastValueBranchInst>(I);
+
+  if (!Inst)
+    return nullptr;
+
+  auto LoweredSourceType = Inst->getOperand()->getType();
+  auto LoweredTargetType = Inst->getCastType();
+  auto SourceType = LoweredSourceType.getSwiftRValueType();
+  auto TargetType = LoweredTargetType.getSwiftRValueType();
+  auto Loc = Inst->getLoc();
+  auto *SuccessBB = Inst->getSuccessBB();
+  auto *FailureBB = Inst->getFailureBB();
+  auto Op = Inst->getOperand();
+  auto &Mod = Inst->getModule();
+  bool isSourceTypeExact = isa<MetatypeInst>(Op);
+
+  // Check if we can statically predict the outcome of the cast.
+  auto Feasibility = classifyDynamicCast(Mod.getSwiftModule(), SourceType,
+                                         TargetType, isSourceTypeExact);
+
+  if (Feasibility == DynamicCastFeasibility::MaySucceed) {
+    return nullptr;
+  }
+
+  SILBuilderWithScope Builder(Inst);
+
+  if (Feasibility == DynamicCastFeasibility::WillFail) {
+    auto *NewI = Builder.createBranch(Loc, FailureBB);
+    EraseInstAction(Inst);
+    WillFailAction();
+    return NewI;
+  }
+
+  // Casting will succeed.
+
+  // Replace by unconditional_cast, followed by a branch.
+  // The unconditional_cast can be skipped, if the result of a cast
+  // is not used afterwards.
+  bool ResultNotUsed = SuccessBB->getArgument(0)->use_empty();
+  SILValue CastedValue;
+  if (Op->getType() != LoweredTargetType) {
+    if (!ResultNotUsed) {
+      auto Src = Inst->getOperand();
+      auto Dest = SILValue();
+      // To apply the bridged casts optimizations.
+      auto BridgedI = optimizeBridgedCasts(
+          Inst, CastConsumptionKind::CopyOnSuccess, false, Src, Dest,
+          SourceType, TargetType, nullptr, nullptr);
+
+      if (BridgedI) {
+        CastedValue = BridgedI;
+      } else {
+        if (!canUseScalarCheckedCastInstructions(Mod, SourceType, TargetType))
+          return nullptr;
+
+        CastedValue = emitSuccessfulScalarUnconditionalCast(
+            Builder, Mod.getSwiftModule(), Loc, Op, LoweredTargetType,
+            SourceType, TargetType, Inst);
+      }
+
+      if (!CastedValue)
+        CastedValue = Builder.createUnconditionalCheckedCastOpaque(
+            Loc, Op, LoweredTargetType);
+    } else {
+      CastedValue = SILUndef::get(LoweredTargetType, Mod);
+    }
+  } else {
+    // No need to cast.
+    CastedValue = Op;
+  }
+
+  auto *NewI = Builder.createBranch(Loc, SuccessBB, CastedValue);
+  EraseInstAction(Inst);
+  WillSucceedAction();
+  return NewI;
+}
```

Comment:
There's a TODO as I mentioned - the optimizer code is not done / will not reach this point at the moment. I will add tests when we have SILOptimizer support.